pkg/validator/validator.go

```diff
@@ -15,12 +15,141 @@
 package validator
 
 import (
+	"bytes"
 	"context"
+	"encoding/json"
 	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"time"
 )
 
-type Validator struct{}
+const (
+	jiraResponseSizeLimitInBytes = 4_000_000 // 4mb
+)
+
+type Validator struct {
+	// Baseurl of the jira API
+	// Example https://your-domain.atlassian.net/rest/api/3
+	baseURL string
+
+	// httpClient is an HTTP client used for making outbound requests.
+	httpClient *http.Client
+}
+
+type JiraIssue struct {
+	Key string `json:"key"`
+	ID  string `json:"id"`
+}
+
+type Match struct {
+	MatchedIssues []int    `json:"matchedIssues"`
+	Errors        []string `json:"errors"`
+}
+
+type MatchResult struct {
+	Matches []Match `json:"matches"`
+}
+
+type matchData struct {
+	IssueIds []string `json:"issueIds"`
+	Jqls     []string `json:"jqls"`
+}
+
+func NewValidator(url string) (*Validator, error) {
+	return &Validator{
+		baseURL:    url,
+		httpClient: &http.Client{Timeout: 10 * time.Second},
+	}, nil
+}
+
+func (v *Validator) MatchIssue(ctx context.Context, issueKey, jql, jiraAccount, apiToken string) (*MatchResult, error) {
+	issue, err := v.jiraIssue(ctx, issueKey, jiraAccount, apiToken)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get jira issue %s: %w", issueKey, err)
+	}
+
+	result, err := v.matchJQL(ctx, issue, jql, jiraAccount, apiToken)
+	if err != nil {
+		return nil, fmt.Errorf("failed to validate jira issue %s: %w", issueKey, err)
+	}
+	return result, nil
+}
+
+func (v *Validator) jiraIssue(ctx context.Context, issueIDOrKey, jiraAccount, apiToken string) (*JiraIssue, error) {
+	// Construct get issue api.
+	// https://developer.atlassian.com/cloud/jira/platform/rest/v3/api-group-issues/#api-rest-api-3-issue-issueidorkey-get
+	u, err := url.Parse(fmt.Sprintf("%s/issue/%s?fields=key,id", v.baseURL, issueIDOrKey))
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse getIssue url: %w", err)
+	}
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, u.String(), nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to construct request: %w", err)
+	}
+
+	header := map[string]string{"Accept": "application/json"}
+
+	var jiraIssue JiraIssue
+	if err := v.makeRequest(req, jiraAccount, apiToken, header, &jiraIssue); err != nil {
+		return nil, err
+	}
+
+	return &jiraIssue, nil
+}
+
+func (v *Validator) matchJQL(ctx context.Context, issue *JiraIssue, jql, jiraAccount, apiToken string) (*MatchResult, error) {
+	// Construct match api.
+	// https://developer.atlassian.com/cloud/jira/platform/rest/v3/api-group-issue-search/#api-rest-api-3-jql-match-post
+
+	u, err := url.Parse(fmt.Sprintf("%s/jql/match", v.baseURL))
```

Comment:
Similar to Seth's example of URL, we should construct a URL object.

Reply:
Done